sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -0,0 +1,287 @@
+
+    /// Creates a new `Primitive` protocol updater from `options`.
+    pub fn new(options: &BootstrapOptions) -> Self {
+        Self {
+            inter_frame_delay: options.inter_frame_delay.unwrap_or(Self::INTER_FRAME_DELAY),
+            flash_erase_delay: options.flash_erase_delay.unwrap_or(Self::FLASH_ERASE_DELAY),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum LegacyBootstrapError {
+    #[error("Boot rom not ready")]
+    NotReady(),
+    #[error("Unknown boot rom error: {0}")]
+    Unknown(u8),
+    #[error("Boot rom error: NOREQUEST")]
+    NOREQUEST(),
+    #[error("Boot rom error: NOMAGIC")]
+    NOMAGIC(),
+    #[error("Boot rom error: TOOBIG")]
+    TOOBIG(),
+    #[error("Boot rom error: TOOHIGH")]
+    TOOHIGH(),
+    #[error("Boot rom error: NOALIGN")]
+    NOALIGN(),
+    #[error("Boot rom error: NOROUND")]
+    NOROUND(),
+    #[error("Boot rom error: BADKEY")]
+    BADKEY(),
+    #[error("Boot rom error: BADSTART")]
+    BADSTART(),
+    #[error("Boot rom error: NOWIPE")]
+    NOWIPE(),
+    #[error("Boot rom error: NOWIPE0")]
+    NOWIPE0(),
+    #[error("Boot rom error: NOWIPE1")]
+    NOWIPE1(),
+    #[error("Boot rom error: NOTEMPTY")]
+    NOTEMPTY(),
+    #[error("Boot rom error: NOWRITE")]
+    NOWRITE(),
+    #[error("Boot rom error: BADADR")]
+    BADADR(),
+    #[error("Boot rom error: OVERFLOW")]
+    OVERFLOW(),
+}
+
+impl From<u8> for LegacyBootstrapError {
+    fn from (value: u8) -> LegacyBootstrapError {
+        match value {
+            // All zeroes or all ones means that the bootloader is not yet ready to respond.
+            0 | 255 => LegacyBootstrapError::NotReady(),
+            // Other values represent particular errors.
+            1 => LegacyBootstrapError::NOREQUEST(),
+            2 => LegacyBootstrapError::NOMAGIC(),
+            3 => LegacyBootstrapError::TOOBIG(),
+            4 => LegacyBootstrapError::TOOHIGH(),
+            5 => LegacyBootstrapError::NOALIGN(),
+            6 => LegacyBootstrapError::NOROUND(),
+            7 => LegacyBootstrapError::BADKEY(),
+            8 => LegacyBootstrapError::BADSTART(),
+            10 => LegacyBootstrapError::NOWIPE(),
+            11 => LegacyBootstrapError::NOWIPE0(),
+            12 => LegacyBootstrapError::NOWIPE1(),
+            13 => LegacyBootstrapError::NOTEMPTY(),
+            14 => LegacyBootstrapError::NOWRITE(),
+            15 => LegacyBootstrapError::BADADR(),
+            16 => LegacyBootstrapError::OVERFLOW(),
+            n => LegacyBootstrapError::Unknown(n),
+        }
+    }
+}
+
+impl UpdateProtocol for Legacy {
+    /// Performs the update protocol using the `transport` with the firmware `payload`.
+    fn update(&self, spi: &dyn Target, payload: &[u8]) -> Result<()> {
+        let frames = Frame::from_payload(payload);
+
+        let mut i = 0;
+        while i < frames.len() {
+            let frame = &frames[i];
+            eprint!("{}.", i);
```

Comment:
This is fine for now, but we should add a TODO to refactor the bootstrap module to take a progressbar callback function (as a follow-up PR).

The SpiFlash module has this; we can move some of the progressbar setup out of opentitantooll/command/spiflash.rs into a more neutral location and reuse it.

Reply:
Done